### PR TITLE
[SPARK-19419] [SPARK-19420] Fix the cross join detection

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
There are two issues in the existing detection of cartesian products. 

1) When users use the outer joins where both sides of the tables are unable to be broadcasted, Spark will still select `BroadcastNestedLoopJoin`. CROSS JOIN syntax is unable to cover the scenario of outer join, but we still issue the following error message:
```
Use the CROSS JOIN syntax to allow cartesian products between these relations
```

2) The existing detection is unable to cover all the cartesian product cases. For example, 
  - Case 1) only having non-equal predicates in join conditions for an inner join. e.g., `tab1.a < tab2.b`
  - Case 2) equi-join's key columns are not sortable and both sides are not small enough for broadcasting and shuffling. That means, it is not qualified for either broadcast hash/shuffle join or sort merge joins.

This PR is to move the cross-join detection back to `BroadcastNestedLoopJoinExec` and `CartesianProductExec`. 

### How was this patch tested?
Added the extra test cases.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 1]
```diff
@@ -215,18 +215,36 @@ class JoinSuite extends QueryTest with SharedSQLContext {
           Row(1, null, 2, 2) ::
           Row(2, 2, 1, null) ::
           Row(2, 2, 2, 2) :: Nil)
+
+      checkAnswer(
+        testData3.as("x").join(testData3.as("y"), $"x.a" > $"y.a"),
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: This is a typical example of cartesian product, but our current detection is unable to find.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 2]
```diff
@@ -584,24 +602,37 @@ class JoinSuite extends QueryTest with SharedSQLContext {
     val cartesianQueries = Seq(
       /** The following should error out since there is no explicit cross join */
       "SELECT * FROM testData inner join testData2",
-      "SELECT * FROM testData left outer join testData2",
-      "SELECT * FROM testData right outer join testData2",
-      "SELECT * FROM testData full outer join testData2",
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: These three queries are examples for outer joins that are unable to be replaced by cross join.

File: sql/core/src/test/resources/sql-tests/inputs/cross-join.sql [hunk 3]
```diff
@@ -33,3 +33,5 @@ create temporary view D(d, vd) as select * from nt1;
 -- Allowed since cross join with C is explicit
 select * from ((A join B on (a = b)) cross join C) join D on (a = d);
 
+-- Cross joins with non-equal predicates
+SELECT * FROM nt1 CROSS JOIN nt2 ON (nt1.k > nt2.k);
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/cross-join.sql]: So far, the SQL syntax allows users to specify the join condition. 

The Dataset API does not allow users to do it, but users still can do it by using the filter, which will be pushed into the cross join.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 4]
```diff
@@ -339,6 +340,18 @@ case class BroadcastNestedLoopJoinExec(
     )
   }
 
+  protected override def doPrepare(): Unit = {
+    if (!withinBroadcastThreshold && !sqlContext.conf.crossJoinEnabled) {
+      throw new AnalysisException(
+        s"""
+           |Both sides of this join are outside the broadcasting threshold and
```
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: should it
- say a cross join is being considered (if it is)
- reference the broadcast threshold to allow the user to increase it?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: I am not sure whether increasing `broadcast threshold ` is a good option to users.

**Comment by gatorsmile** [comment]
Let me first push the new changes to resolve the SparkR issue. However, I still need time to add the corresponding test cases to Scala.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 5]
```diff
@@ -215,18 +215,61 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
+      var e = intercept[AnalysisException](testData3.join(testData3).collect())
+      assert(e.getMessage.contains(msg))
+      e = intercept[AnalysisException] {
+        testData3.as("x").join(testData3.as("y"), $"x.a" > $"y.a").collect()
+      }
+      assert(e.getMessage.contains(msg))
+    }
+  }
+
+  test("subquery converted to cartesian product join") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: This is another typical example for cartesian product, which is converted from subquery.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @srinathshankar @sameeragarwal @hvanhovell @cloud-fan

**Comment by cloud-fan** [comment]
is CROSS JOIN a logical or physical concept?

**Comment by gatorsmile** [comment]
Cross join is a physical concept. In Spark 2.0, we detected it like what this PR did. In Spark 2.1, we moved it to Optimizer. Basically, this PR is to change it back.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 6]
```diff
@@ -339,6 +340,33 @@ case class BroadcastNestedLoopJoinExec(
...
+                 |${right.treeString(false).trim}
+                 |Join condition is missing or trivial.
+                 |Use the CROSS JOIN syntax to allow cartesian products between these relations.
+               """.stripMargin)
+          }
+        case _ if !withinBroadcastThreshold || condition.isEmpty =>
+          throw new AnalysisException(
+            s"""Both sides of this join are outside the broadcasting threshold and
+               |computing it could be prohibitively expensive. To explicitly enable it,
+               |Please set ${SQLConf.CROSS_JOINS_ENABLED.key} = true.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: shall we support `CROSS OUTER JOIN`, `CROSS LEFT JOIN`, etc?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 7]
```diff
@@ -339,6 +340,33 @@ case class BroadcastNestedLoopJoinExec(
     )
   }
 
+  protected override def doPrepare(): Unit = {
+    if (!sqlContext.conf.crossJoinEnabled) {
+      joinType match {
+        case Cross => // Do nothing
+        case Inner =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: can you add comments to say when we will hit this branch?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 8]
```diff
@@ -213,7 +213,12 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           }
         // This join could be very slow or OOM
         joins.BroadcastNestedLoopJoinExec(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: when we broadcast, is it still a cross join?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: : ) This is also my question.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: Sure

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: That depends on how we defined it. See the original PR: https://github.com/apache/spark/pull/13209

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: cc @sameeragarwal @srinathshankar

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 9]
```diff
@@ -584,24 +602,39 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
 
     cartesianQueries.foreach(checkCartesianDetection)
+
+    val broadcastNestedLoopJoinOutOfBroadcastThresholdQueries = Seq(
+      /** The following should error out since the flag spark.sql.crossJoin.enabled is not on */
+      "SELECT * FROM testData left outer join testData2",
+      "SELECT * FROM testData right outer join testData2",
+      "SELECT * FROM testData full outer join testData2",
+      "SELECT * FROM testData full outer join testData2 on testData.key > testData2.a",
+      "SELECT b FROM testData2 WHERE EXISTS(SELECT a FROM testData3)"
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: This is a subquery that will use `BroadcastNestedLoopJoin`.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 10]
```diff
@@ -584,24 +602,39 @@ class JoinSuite extends QueryTest with SharedSQLContext {
...
       "SELECT * FROM testData, testData2",
       "SELECT * FROM testData, testData2 where testData.key = 1 and testData2.a = 22",
+      "SELECT * FROM testData, testData2 where testData.key > testData2.a",
+      "SELECT * FROM testData INNER JOIN testData2 ON testData.key + testData2.a = testData2.b",
       /** The following should fail because after reordering there are cartesian products */
       "select * from (A join B on (A.key = B.key)) join D on (A.key=D.a) join C",
       "select * from ((A join B on (A.key = B.key)) join C) join D on (A.key = D.a)",
       /** Cartesian product involving C, which is not involved in a CROSS join */
-      "select * from ((A join B on (A.key = B.key)) cross join D) join C on (A.key = D.a)");
+      "select * from ((A join B on (A.key = B.key)) cross join D) join C on (A.key = D.a)")
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: The estimated size of `testData`, `testData2`, and `testData3` are much larger than the broadcast threshold. Thus, it does not consider the cases when the tables are small enough to broadcast. Thus, when the tables are small enough to broadcast, should we still block users? In that case, the join is not expensive.

It affects the whole detection logics rewriting. Thanks! @rxin @hvanhovell @cloud-fan @sameeragarwal

**Pull request closed without merging**